File: ts/messaging.ts
import { EventEmitter } from "./events";
import { concatBytes, createChunkId, frameChunkId, normalizeMessageData, parseChunkId, uuid7 } from "./utils";

type Mode = "text" | "binary";

type JsonRpcRequest = {
  id: string | number | null;
  method: string;
  params: Record<string, unknown>;
  jsonrpc: "2.0";
};

type JsonRpcResponse = {
  id: string | number;
  result: Record<string, unknown> | null;
  jsonrpc: "2.0";
};

type JsonRpcError = {
  id: string | number;
  error: {
    code: number;
    message: string;
    data?: Record<string, unknown> | null;
  };
  jsonrpc: "2.0";
};

type MessageHandlers = {
  start_session: [sessionId: string, byteLength: number, sha256: string, mode: Mode];
  request_chunk: [sessionId: string, chunkId: bigint, byteOffset: number, byteLength: number];
  chunk_content: [chunkId: bigint, data: Uint8Array];
  timeout: [sessionId: string];
};

type MessageHandler = (...args: any[]) => Promise<unknown> | unknown;

export class DekaiDataChannelMessaging extends EventEmitter<MessageHandlers> {
  private readonly channel: RTCDataChannel;
  private readonly futures = new Map<
    string | number,
    {
      resolve: (
        value:
          | Record<string, unknown>
          | PromiseLike<Record<string, unknown> | null>
          | null,
      ) => void;
      reject: (reason?: unknown) => void;
    }
  >();
  private readonly textDecoder = new TextDecoder();

  constructor(channel: RTCDataChannel) {
    super();
    this.channel = channel;
    this.channel.addEventListener("message", (event) => {
      void this.onMessage(event.data as string | Blob | ArrayBuffer | ArrayBufferView);
    });
  }

  call(method: "start_session", params: { session_id: string; byte_length: number; sha256: string; mode: Mode }): Promise<Record<string, unknown> | null>;
  call(method: "request_chunk", params: { session_id: string; chunk_id: bigint; byte_offset: number; byte_length: number }): Promise<Record<string, unknown> | null>;
  call(method: string, params: Record<string, unknown>): Promise<Record<string, unknown> | null> {
    const requestId = uuid7();
    const payload = this.encodeJson({
      id: requestId,
      method,
      params: this.serializeParams(params),
      jsonrpc: "2.0",
    });

    const promise = new Promise<Record<string, unknown> | null>((resolve, reject) => {
      this.futures.set(requestId, { resolve, reject });
    });
    this.channel.send(payload);
    return promise;
  }

  notify(method: "timeout", params: { session_id: string }): void;
  notify(method: "request_chunk", params: { session_id: string; chunk_id: bigint; byte_offset: number; byte_length: number }): void;
  notify(method: string, params: Record<string, unknown>): void {
    const payload = this.encodeJson({
      id: null,
      method,
      params: this.serializeParams(params),
      jsonrpc: "2.0",
    });
    this.channel.send(payload);
  }

  sendChunk(chunkId: bigint, data: Uint8Array): void {
    const message = concatBytes([frameChunkId(chunkId), data]);
    const normalized = new Uint8Array(message.byteLength);
    normalized.set(message);
    this.channel.send(normalized);
  }

  private async onMessage(rawData: string | Blob | ArrayBuffer | ArrayBufferView): Promise<void> {
    const data = await normalizeMessageData(rawData);
    if (typeof data === "string") {
      this.processJsonMessage(data);
      return;
    }

    if (data[0] === 0x7b) {
      this.processJsonMessage(this.textDecoder.decode(data));
      return;
    }

    if (data[0] === 0x66) {
      const chunkId = parseChunkId(data.subarray(0, 8));
      const chunk = data.subarray(8);
      await this.emitAsync("chunk_content", chunkId, chunk);
    }
  }

  private processJsonMessage(text: string): void {
    const message = JSON.parse(text) as JsonRpcRequest | JsonRpcResponse | JsonRpcError;
    if ("method" in message) {
      void this.processRequest(message);
      return;
    }

    const future = this.futures.get(message.id);
    if (future === undefined) {
      return;
    }
    this.futures.delete(message.id);

    if ("error" in message) {
      future.reject(new Error(message.error.message));
      return;
    }

    future.resolve(message.result ?? null);
  }

  private async processRequest(request: JsonRpcRequest): Promise<void> {
    try {
      const result = await this.dispatchRequest(request.method, request.params);
      if (request.id === null) {
        return;
      }

      this.channel.send(this.encodeJson({
        id: request.id,
        result: (result as Record<string, unknown> | null) ?? null,
        jsonrpc: "2.0",
      }));
    } catch (error) {
      if (request.id === null) {
        return;
      }

      this.channel.send(this.encodeJson({
        id: request.id,
        error: {
          code: -32000,
          message: error instanceof Error ? error.message : String(error),
          data: null,
        },
        jsonrpc: "2.0",
      }));
    }
  }

  private dispatchRequest(method: string, params: Record<string, unknown>): Promise<unknown> {
    switch (method) {
      case "start_session":
        return this.emitAsync(
          "start_session",
          String(params.session_id),
          Number(params.byte_length),
          String(params.sha256),
          params.mode as Mode,
        );
      case "request_chunk":
        return this.emitAsync(
          "request_chunk",
          String(params.session_id),
          this.normalizeChunkId(params.chunk_id),
          Number(params.byte_offset),
          Number(params.byte_length),
        );
      case "timeout":
        return this.emitAsync("timeout", String(params.session_id));
      default:
        throw new Error(`Unknown method: ${method}`);
    }
  }

  private async emitAsync<EventName extends keyof MessageHandlers>(
    event: EventName,
    ...args: MessageHandlers[EventName]
  ): Promise<unknown> {
    let result: unknown = null;
    const eventHandlers = this.getHandlers(event) as Set<MessageHandler> | undefined;
    if (eventHandlers === undefined || eventHandlers.size === 0) {
      return result;
    }

    for (const handler of eventHandlers) {
      result = await handler(...args);
    }
    return result;
  }

  private encodeJson(payload: JsonRpcRequest | JsonRpcResponse | JsonRpcError): string {
    return JSON.stringify(payload);
  }

  private serializeParams(params: Record<string, unknown>): Record<string, unknown> {
    const serialized: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(params)) {
      serialized[key] = typeof value === "bigint" ? value.toString() : value;
    }
    return serialized;
  }

  private normalizeChunkId(value: unknown): bigint {
    if (typeof value === "bigint") {
      return value;
    }
    if (typeof value === "number") {
      return BigInt(value);
    }
    if (typeof value === "string") {
      return BigInt(value);
    }
    return createChunkId();
  }
}

export type { Mode };

File: dekai_datachannel/_messaging.py
from aiortc import RTCDataChannel
from logging import getLogger
from uuid import UUID, uuid7
from typing import Literal, overload, Protocol
from asyncio import Future
from dataclasses import dataclass
from pydantic import TypeAdapter
import asyncio

logger = getLogger("dekai_datachannel")

CHUNK_ID_MASK = 0x000FFFFFFFFFFFFF


@dataclass
class JSONRPCRequest:
    id: str | int | None
    method: str
    params: dict
    jsonrpc: Literal["2.0"] = "2.0"


_jsonrpc_request_adapter = TypeAdapter(JSONRPCRequest)


@dataclass
class JSONRPCErrorBody:
    code: int
    message: str
    data: dict | None = None


@dataclass
class JSONRPCError:
    id: str | int
    error: JSONRPCErrorBody
    jsonrpc: Literal["2.0"] = "2.0"


@dataclass
class JSONRPCResponse:
    id: str | int
    result: dict | None = None
    jsonrpc: Literal["2.0"] = "2.0"


type JSONRPCMessages = JSONRPCRequest | JSONRPCResponse | JSONRPCError

_jsonrpc_messages_adapter = TypeAdapter(JSONRPCMessages)


class StartSessionHandler(Protocol):
    async def __call__(
        self,
        session_id: UUID,
        byte_length: int,
        sha256: str,
        mode: Literal["text", "binary"],
    ) -> None: ...


class RequestChunkHandler(Protocol):
    async def __call__(
        self,
        session_id: UUID,
        chunk_id: int,
        byte_offset: int,
        byte_length: int,
    ) -> None: ...


class ChunkContentHandler(Protocol):
    async def __call__(
        self,
        chunk_id: int,
        data: bytes,
    ) -> None: ...


class TimeoutHandler(Protocol):
    async def __call__(
        self,
        session_id: UUID,
    ) -> None: ...


type MessageHandlers = (
    StartSessionHandler | RequestChunkHandler | ChunkContentHandler | TimeoutHandler
)


class DekaiDataChannelMessaging:
    def __init__(self, channel: RTCDataChannel):
        self.__channel = channel
        self.__handlers: dict[str, MessageHandlers] = {}

        channel.on("message", self._on_message)
        channel.on("open", self._on_open)
        channel.on("close", self._on_close)
        channel.on("error", self._on_error)

        self.__futures: dict[str | int, Future] = {}

    @property
    def channel(self) -> RTCDataChannel:
        return self.__channel

    def _on_open(self):
        logger.info("Data channel is open")

    def _on_close(self):
        logger.info("Data channel is closed")

    def _on_error(self, error):
        logger.error(f"Data channel error: {error}")

    def _on_message(self, message: str | bytes):
        if isinstance(message, str):
            message = message.encode("utf-8")

        if b"{" == message[0:1]:
            self._process_json_message(message)
        elif b"f" == message[0:1]:
            self._process_frame_message(message)
        else:
            logger.warning(f"Unknown message type: {message}")

    def _process_json_message(self, message: bytes):
        try:
            data = _jsonrpc_messages_adapter.validate_json(message)
            if isinstance(data, JSONRPCResponse):
                call_id = data.id
                if call_id in self.__futures:
                    future = self.__futures.pop(call_id)
                    if not future.done():
                        future.set_result(data.result)
            elif isinstance(data, JSONRPCError):
                call_id = data.id
                if call_id in self.__futures:
                    future = self.__futures.pop(call_id)
                    if not future.done():
                        future.set_exception(Exception(data.error))
            elif isinstance(data, JSONRPCRequest):
                coro = self._process_json_request(data)
                asyncio.create_task(coro)

        except Exception as e:
            # TODO: Response or notify with error message
            logger.error(f"Invalid JSON-RPC message: {e}")
            return

    async def _process_json_request(self, request: JSONRPCRequest):
        if request.id is None:
            try:
                await self._emit(request.method, **request.params)
            except Exception as e:
                logger.error(
                    "JSON-RPC notification handler failed: method=%s params=%s error=%s",
                    request.method,
                    request.params,
                    e,
                    exc_info=e,
                )
        else:
            try:
                result = await self._emit(request.method, **request.params)
            except Exception as e:
                error = JSONRPCError(
                    id=request.id,
                    error=JSONRPCErrorBody(code=-32000, message=str(e)),
                    jsonrpc="2.0",
                )
                self.channel.send(_jsonrpc_messages_adapter.dump_json(error))
            else:
                response = JSONRPCResponse(id=request.id, result=result)
                self.channel.send(_jsonrpc_messages_adapter.dump_json(response))

    def _process_frame_message(self, message: bytes):
        message_id = int.from_bytes(message[0:8], "little")
        message_id = message_id >> 8
        message_body = message[8:]
        coro = self._emit("chunk_content", message_id, message_body)
        asyncio.create_task(coro)

    @overload
    def on(self, method: Literal["start_session"], handler: StartSessionHandler): ...
    @overload
    def on(self, method: Literal["request_chunk"], handler: RequestChunkHandler): ...
    @overload
    def on(self, method: Literal["chunk_content"], handler: ChunkContentHandler): ...
    @overload
    def on(self, method: Literal["timeout"], handler: TimeoutHandler): ...
    def on(self, method: str, handler: MessageHandlers):
        existing = self.__handlers.get(method)
        if existing is not None and existing != handler:
            raise ValueError(f"Handler for method '{method}' already exists")
        self.__handlers[method] = handler

    @overload
    def off(self, method: Literal["start_session"], handler: StartSessionHandler): ...
    @overload
    def off(self, method: Literal["request_chunk"], handler: RequestChunkHandler): ...
    @overload
    def off(self, method: Literal["chunk_content"], handler: ChunkContentHandler): ...
    @overload
    def off(self, method: Literal["timeout"], handler: TimeoutHandler): ...
    def off(self, method: str, handler: MessageHandlers):
        if method in self.__handlers:
            existing = self.__handlers[method]
            if existing == handler:
                del self.__handlers[method]
            else:
                raise ValueError(
                    f"Handler for method '{method}' does not match the provided handler"
                )

    async def _emit(self, method: str, *args, **kwargs):
        if method in self.__handlers:
            handler = self.__handlers[method]
            await handler(*args, **kwargs)

    @overload
    def call(
        self,
        method: Literal["start_session"],
        *,
        session_id: UUID,
        byte_length: int,
        sha256: str,
        mode: Literal["text", "binary"],
    ) -> Future: ...

    @overload
    def call(
        self,
        method: Literal["request_chunk"],
        *,
        session_id: UUID,
        chunk_id: int,
        byte_offset: int,
        byte_length: int,
    ) -> Future: ...

    def call(self, method: str, **kwargs) -> Future:
        request_id = str(uuid7())
        request = JSONRPCRequest(
            id=request_id,
            method=method,
            params=kwargs,
        )
        future = Future()
        self.__futures[request_id] = future
        self.channel.send(_jsonrpc_request_adapter.dump_json(request))
        return future

    @overload
    def notify(
        self,
        method: Literal["start_session"],
        *,
        session_id: UUID,
        byte_length: int,
        sha256: str,
        mode: Literal["text", "binary"],
    ) -> None: ...

    @overload
    def notify(
        self,
        method: Literal["request_chunk"],
        *,
        session_id: UUID,
        chunk_id: int,
        byte_offset: int,
        byte_length: int,
    ) -> None: ...

    @overload
    def notify(
        self,
        method: Literal["timeout"],
        *,
        session_id: UUID,
    ) -> None: ...

    def notify(self, method: str, **kwargs) -> None:
        request = JSONRPCRequest(
            id=None,
            method=method,
            params=kwargs,
        )
        self.channel.send(_jsonrpc_request_adapter.dump_json(request))

    async def send_chunk(
        self,
        chunk_id: int,
        data: bytes,
    ):
        if chunk_id < 0 or chunk_id > CHUNK_ID_MASK:
            raise ValueError("chunk_id must fit within 52 bits.")

        frame_id = (chunk_id << 8) | 0x66
        chunk_id_bytes = frame_id.to_bytes(8, "little")
        message = chunk_id_bytes + data
        self.channel.send(message)
